internal/datasource/host/info.go

```diff
@@ -0,0 +1,348 @@
+		}
+
+		return containsContainerReference(i.selfCgroupLocation), nil
+	})
+
+	if err != nil {
+		slog.Warn("Unable to determine if resource is a container or not", "error", err)
+		return false
+	}
+
+	if result, ok := res.(bool); ok {
+		return result
+	}
+
+	return false
+}
+
+func (i *Info) GetContainerInfo() *v1.Resource_ContainerInfo {
+	return &v1.Resource_ContainerInfo{
+		ContainerInfo: &v1.ContainerInfo{
+			Id: i.getContainerID(),
+		},
+	}
+}
+
+func (i *Info) GetHostInfo(ctx context.Context) *v1.Resource_HostInfo {
+	hostname, err := i.exec.GetHostname()
+	if err != nil {
+		slog.WarnContext(ctx, "Unable to get hostname", "error", err)
+	}
+
+	return &v1.Resource_HostInfo{
+		HostInfo: &v1.HostInfo{
+			Id:          i.getHostID(ctx),
+			Hostname:    hostname,
+			ReleaseInfo: i.getReleaseInfo(ctx, i.osReleaseLocation),
+		},
+	}
+}
+
+func containsContainerReference(cgroupFile string) bool {
+	data, err := os.ReadFile(cgroupFile)
+	if err != nil {
+		slog.Error("Unable to check if cgroup file contains a container reference", "error", err)
+		return false
+	}
+
+	scanner := bufio.NewScanner(strings.NewReader(string(data)))
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+		if strings.Contains(line, k8sKind) || strings.Contains(line, docker) || strings.Contains(line, conatinerd) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (i *Info) getContainerID() string {
+	res, err, _ := singleflightGroup.Do(GetContainerIDKey, func() (interface{}, error) {
+		containerID, err := getContainerIDFromMountInfo(i.mountInfoLocation)
+		return uuid.NewMD5(uuid.NameSpaceDNS, []byte(containerID)).String(), err
+	})
+
+	if err != nil {
+		slog.Error("Could not get container ID", "error", err)
+		return ""
+	}
+
+	if result, ok := res.(string); ok {
+		return result
+	}
+
+	return ""
+}
+
+// getContainerID returns the container ID of the current running environment.
+// Supports cgroup v1 and v2. Reading "/proc/1/cpuset" would only work for cgroups v1
+// mountInfo is the path: "/proc/self/mountinfo"
+func getContainerIDFromMountInfo(mountInfo string) (string, error) {
+	mInfoFile, err := os.Open(mountInfo)
+	if err != nil {
+		return "", fmt.Errorf("could not read %s: %w", mountInfo, err)
+	}
+	defer func(f *os.File, fileName string) {
+		closeErr := f.Close()
+		if closeErr != nil {
+			slog.Error("Unable to close file", "file", fileName, "error", closeErr)
+		}
+	}(mInfoFile, mountInfo)
+
+	fileScanner := bufio.NewScanner(mInfoFile)
+	fileScanner.Split(bufio.ScanLines)
+
+	var lines []string
+	for fileScanner.Scan() {
+		lines = append(lines, fileScanner.Text())
+	}
+
+	for _, line := range lines {
+		splitLine := strings.Split(line, " ")
+		for _, word := range splitLine {
+			containerID := getContainerIDFromPatterns(word)
+			if containerID != "" {
+				return containerID, nil
+			}
+		}
+	}
+
+	return "", fmt.Errorf("container ID not found in %s", mountInfo)
+}
+
+func getContainerIDFromPatterns(word string) string {
```

Comment:
Would it be a good idea to add an example of a match for each pattern as a comment, since interpreting Regex is always a little iffy?